### PR TITLE
HIVE-24504: VectorFileSinkArrowOperator does not serialize complex types correctly

### PR DESCRIPTION
**Pull request by pvary**
### What changes were proposed in this pull request?
Use an empty batch to generate the schema for the empty results

### Why are the changes needed?
Clients expect the full schema even for empty results

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unit and other test

### CODE REVIEW

File: ql/src/test/org/apache/hadoop/hive/ql/io/arrow/TestSerializer.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfoUtils;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.List;
+
+public class TestSerializer {
+  @Test
+  public void testEmptyArray() {
```
Comment (pgaref) [inline ql/src/test/org/apache/hadoop/hive/ql/io/arrow/TestSerializer.java:31]: Nit: Would probably name it testEmptyList for consistency with the Serializer

File: ql/src/test/org/apache/hadoop/hive/ql/io/arrow/TestSerializer.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    List<TypeInfo> typeInfos = TypeInfoUtils.getTypeInfosFromTypeString(
+        "struct<b:array<tinyint>,c:map<string,string>,d:struct<e:array<tinyint>,f:map<string,string>>>");
+    List<String> fieldNames = Arrays.asList(new String[] { "a" });
+    Serializer converter = new Serializer(new HiveConf(), "attemptId", typeInfos, fieldNames);
+    ArrowWrapperWritable writable = converter.emptyBatch();
+    Assert.assertEquals(
+        "Schema<a: Struct<b: List<$data$: Int(8, true)>, c: List<$data$: Struct<keys: Utf8, values: Utf8>>, " +
+            "d: Struct<e: List<$data$: Int(8, true)>, f: List<$data$: Struct<keys: Utf8, values: Utf8>>>>>",
+        writable.getVectorSchemaRoot().getSchema().toString());
+  }
```
Comment (pgaref) [inline ql/src/test/org/apache/hadoop/hive/ql/io/arrow/TestSerializer.java:71]: Would it make sense to test nested types here as well? like list<list> ?

File: ql/src/test/org/apache/hadoop/hive/ql/io/arrow/TestSerializer.java (hunk 1 above)
Comment (pvary) [reply ql/src/test/org/apache/hadoop/hive/ql/io/arrow/TestSerializer.java]: The name was based on the Hive type,  but I think both makes sense so renamed 😄

File: ql/src/test/org/apache/hadoop/hive/ql/io/arrow/TestSerializer.java (hunk 2 above)
Comment (pvary) [reply ql/src/test/org/apache/hadoop/hive/ql/io/arrow/TestSerializer.java:71]: Addes some more tests to cover every nested type at least once

File: ql/src/test/org/apache/hadoop/hive/ql/io/arrow/TestSerializer.java (hunk 1 above)
Comment (pgaref) [reply ql/src/test/org/apache/hadoop/hive/ql/io/arrow/TestSerializer.java]: > The name was based on the Hive type, but I think both makes sense so renamed 😄

Cant disagree with that :D

**Review by pgaref — APPROVED** [review]
Thanks @pvary ! Changes LGTM -- tests pending

**Review by szlta — APPROVED** [review]

**Pull request merged**